### PR TITLE
Create transaction's route, model and service with functions

### PR DESCRIPTION
**Pull request by mgoniszewska**

### CODE REVIEW

File: src/models/transaction.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const mongoose = require('mongoose');
+
+const transactionSchema = new mongoose.Schema({
+    user: mongoose.Types.ObjectId,
+    owner: mongoose.Types.ObjectId,
+    house: mongoose.Types.ObjectId,
+    timestamp: Date,
```
Comment (ike-tom) [inline src/models/transaction.js:7]: you can add default Date.now

File: src/controllers/transactions.controllers.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    if (response.status === 'invalid') {
+      return res.status(StatusCodes.BAD_REQUEST).json(response);
+    }
+
+    return res.status(StatusCodes.CREATED).json(response);
+  });
+};
+
+return transactionControllers;
```
Comment (ike-tom) [inline src/controllers/transactions.controllers.js:21]: shouldn't there be export default transactionControllers?

**Review by ike-tom — CHANGES_REQUESTED** [review]
please take a look at my comments. Thank you :)

File: src/controllers/transactions.controllers.js (hunk 2 above)
Comment (mgoniszewska) [reply src/controllers/transactions.controllers.js:21]: Yes, you're right :)

**Review by ike-tom — APPROVED** [review]

File: src/controllers/user.controllers.js [hunk 3]
```diff
@@ -33,8 +33,32 @@ export const userLogin = async (data) => {
...
+  const user = await User.findById(id).exec();
+
+  if (!user || !user._id) {
+    return { status: 'invalid', message: 'There is no user with this id.' };
+  }
+
+  return { status: 'success', user };
+};
+
+export const getPayment = async (id, price) => {
```
Comment (FilipHalon) [inline src/controllers/user.controllers.js:52]: How about `deductPayment` or `collectPayment`? The name `getPayment` could suggest the function only fetches unchanged payment.

File: src/controllers/transaction.controllers.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import transactionModel from '../models/transaction';
+import userControllers from './user.controllers';
+import { getById as getHouseById, changeOwner as changeHouseOwner } from './house.controllers';
+
+const addTransaction = async (buyerId, sellerId, houseId, price) => {
+  console.log({ buyerId, sellerId, houseId, price });
```
Comment (FilipHalon) [inline src/controllers/transaction.controllers.js:6]: I guess it could be removed.

**Review by FilipHalon — CHANGES_REQUESTED** [review]

File: src/controllers/transaction.controllers.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    return { status: 'invalid', message: 'Transaction failed.' };
+  }
+
+  if (buyer.cash < price) {
+    return { status: 'invalid', message: 'Not enough cash.' };
+  }
+
+  await userControllers.collectPayment(buyer._id, price);
+  await userControllers.addCash(seller._id, price);
+  await changeHouseOwner(houseId, buyer._id);
```
Comment (FilipHalon) [inline src/controllers/transaction.controllers.js:38]: It seems to be possible that the same user is both a buyer and a seller.

File: src/controllers/transaction.controllers.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  return transaction;
+};
+
+export const buyHouse = async (data) => {
+  const { buyerId, houseId, price } = data;
+
+  let { data: buyer } = await userControllers.getById(buyerId);
+  let { data: house } = await getHouseById(houseId);
+  let { data: seller } = await userControllers.getById(house.owner);
```
Comment (FilipHalon) [inline src/controllers/transaction.controllers.js:26]: I'm getting `undefined` for `buyer` and `house` resulting in an error when it comes to this line. `getById` doesn't return anything and the destructurings of the results of the await calls don't expect an object with the structure `{ status: '', message: '' }`.

**Review by FilipHalon — APPROVED** [review]

File: src/controllers/transaction.controllers.js (hunk 6 above)
Comment (FilipHalon) [inline src/controllers/transaction.controllers.js:26]: I'm getting `undefined` for `buyer` and `house` resulting in an error when it comes to this line. `getById` doesn't return anything and the destructurings of the results of the await calls don't expect an object with the structure `{ status: '', message: '' }`.

File: src/controllers/transaction.controllers.js (hunk 5 above)
Comment (FilipHalon) [inline src/controllers/transaction.controllers.js:38]: It seems to be possible that the same user is both a buyer and a seller.

**Review by FilipHalon — CHANGES_REQUESTED** [review]

File: src/controllers/transaction.controllers.js [hunk 7]
```diff
@@ -21,12 +21,26 @@ const addTransaction = async (buyerId, sellerId, houseId, price) => {
 export const buyHouse = async (data) => {
   const { buyerId, houseId, price } = data;
```
Comment (FilipHalon) [inline src/controllers/transaction.controllers.js:22]: `buyerId` and `houseId` don't exist in `data`. Something like this should solve the problem: `const { buyer: buyerId, house: houseId, price } = data;`.

**Review by FilipHalon — CHANGES_REQUESTED** [review]

File: src/controllers/transaction.controllers.js (hunk 7 above)
Comment (mgoniszewska) [reply src/controllers/transaction.controllers.js:22]: It's only used in POST _transactions/_  route and the data parameter is actually req.body.

Comment (FilipHalon) [reply src/controllers/transaction.controllers.js:22]: I see. These are the keys of data:
Screenshot from 2022-03-06 13-52-33
Destructed `buyerId` and `houseId` aren't present in the data so they are `undefined`. What I suggest is to destructure `buyer` and `house` from the data and alias them to `buyerId`, `houseId` respectively.

Comment (mgoniszewska) [reply src/controllers/transaction.controllers.js:22]: image

This data contains only strings with ids of documents necessary to create transaction. It's not representing the data in the database.

Comment (FilipHalon) [reply src/controllers/transaction.controllers.js:22]: I see. My bad. I thought you defined the keys but it was me. I'm sorry for the hassle.

**Review by FilipHalon — APPROVED** [review]

**Pull request merged**